python/hail/methods/misc.py

```diff
@@ -358,4 +368,213 @@ def window_by_locus(mt: MatrixTable, bp_window_size: int) -> MatrixTable:
     :class:`.MatrixTable`
     """
     require_partition_key_locus(mt, 'window_by_locus')
-    return MatrixTable(mt._jvds.windowVariants(bp_window_size))
\ No newline at end of file
+    return MatrixTable(mt._jvds.windowVariants(bp_window_size))
+
+
+@typecheck(a=np.ndarray,
+           radius=oneof(int, float))
+def array_windows(a, radius):
+    """Returns start and stop indices for window around each array value.
+
+    Examples
+    --------
+
+    >>> hl.array_windows(np.array([1, 2, 4, 4, 6, 8]), 2)
+    (array([0, 0, 1, 1, 2, 4]), array([2, 4, 5, 5, 6, 6]))
+
+    >>> hl.array_windows(np.array([-10.0, -2.5, 0.0, 0.0, 1.2, 2.3, 3.0]), 2.5)
+    (array([0, 1, 1, 1, 2, 2, 4]), array([1, 4, 6, 6, 7, 7, 7]))
+
+    Notes
+    -----
+    For a non-decreasing array ``a``, the resulting ``starts`` and ``stops``
+    arrays have the same length as ``a`` and the property that, for all indices
+    ``i``, ``[starts[i], stops[i])`` is the maximal range of indices ``j`` such
+    that ``a[i] - radius <= a[j] <= a[i] + radius``.
+
+    Index ranges are start-inclusive and stop-exclusive. This function is
+    especially useful in conjunction with
+    :meth:`.BlockMatrix.sparsify_row_intervals`.
+
+    Parameters
+    ----------
+    a: :obj:`ndarray` of signed integer or float values
+        Non-decreasing 1-dimensional array of values.
+    radius: :obj:`float`
+        Non-negative radius of window for values.
+
+    Returns
+    -------
+    (:class:`ndarray` of :obj:`int64`, :class:`ndarray` of :obj:`int64`)
+        Tuple of start indices array and stop indices array.
+    """
+    if radius < 0:
+        raise ValueError(f'array_windows: radius must be non-negative, found {radius}')
+    if a.ndim != 1:
+        raise ValueError("array_windows: 'a' must be 1-dimensional")
+    if not (np.issubdtype(a.dtype, np.signedinteger) or np.issubdtype(a.dtype, np.floating)):
+        if None in a:
+            raise ValueError(f"array_windows: 'a' contains a None value")
+        else:
+            raise ValueError(f"array_windows: 'a' must be an ndarray of signed integer or float values, "
+                             f"found dtype {str(a.dtype)}")
+    if (not np.all(a[:-1] <= a[1:])) or (a.size == 1 and np.isnan(a[0])):
```

Comment:
I don't need to check every element for `nan` because `x <= y` evaluates to false if either is `nan`.  Yeah, it's more precise to say`non-decreasing with respect to index`. This means that `i < j => a(i) <= a(j)`, as opposed to `increasing` which means `i < j => a(i) < a(j)`. I've changed it to ascending order here.